### ISSUE TITLE
Is there a way to log the users' chat record?

### ISSUE THREAD

**Issue by winrey** [oragono/oragono#631]
Sorry but I read the docs but find noting about this. (it seems only for errlog)

Am I wrong?

And it would be better if it support log the chat record to a directory and log each different channel to a different file.

**Comment by slingamn** [comment]
You're correct, there's no code support for this right now. In the past I've recommended that people run a logbot, e.g., https://github.com/globau/logbot/.

#348 is for persisting history to a database. However, since our overall philosophy is batteries-included, I'd consider a feature that logs channels to files (probably with a chanserv preference to disable logging?).

**Comment by DanielOaks** [comment]
I intentionally haven't included this functionality so far because of privacy issues. If we were to support this it would only ever be for channels, it would default to off and disabled, and you could only enable it by both editing the config file and enabling some setting in the channel itself (chanserv, etc). If it is enabled on a channel, the server should make the user aware in some way that the channel is being logged when they join it.

**Comment by slingamn** [comment]
Agree 100% with disabled-by-default; apart from any other considerations, the typical server may not have the disk space to log everything.

>If it is enabled on a channel, the server should make the user aware in some way that the channel is being logged when they join it.

I'm not a huge fan of features like this, because of (a) a philosophical belief that FOSS code should "work for" the hardware owner (b) I think it creates a false sense of security among users, because an administrator can just comment out all the warnings and recompile.

**Comment by DanielOaks** [comment]
Hm. Working for the hardware owner is good, so long as the people using the system know what they're getting into imo, since IRC has a really strong anti-logging bias I think it makes sense to notify them in no uncertain terms when that's not the case (see freenode's [notification requirements when channel logging](https://freenode.net/policies#public-logging) for example, they exist because of the expectations IRC users have). The being able to comment out warnings and recompile is a consideration for sure, but one I don't really worry about too much, we can't stop people from being dodgy if they really want to be. Maybe that'd be somewhat avoided by making the notifications not too disruptive. Thoughts on maybe using a custom numeric or message that we create to notify users about it? (maybe a `NOTE` or something would be interesting, if use a certain code clients could integrate the notice into their UI perhaps)

Not super related but there's an issue over here that discusses some things around channel logging done by users, some of the follow-up comments are interesting: https://github.com/ircv3/ircv3-ideas/issues/21

**Comment by slingamn** [comment]
#379 is done now; this can use that same channel preference together with the same server setting structure as persistent history ("mandatory", "opt-out", "opt-in", "disabled").
